File: Graph/Kruskal/kru.cpp
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
  public:
    int findParent(int a,vector<int>& parent){
        if(parent[a] == a)return a;
        return parent[a] = findParent(parent[a],parent);
    }
    bool unionGroup(int a, int b, vector<int>& parent, vector<int>& rank){
        int parB = findParent(b,parent);
        int parA = findParent(a,parent);
        int rankA = rank[a];
        int rankB = rank[b];
        if(parA == parB)return false;
        if(rankA >= rankB){
            parent[parB] = parA;
            rank[parA]++;
        }else{
            parent[parA] = parB;
            rank[parB]++;
        }
        return true;
    }
    static bool cmp(vector<int>& a, vector<int> & b){
        return a[2]<b[2];
    }
    int kruskalsMST(int V, vector<vector<int>> &edges) {
        // code here
        sort(edges.begin(),edges.end(),cmp);
        vector<int> parent(V);
        vector<int>rank(V,1);
        for(int i=0;i<V;i++){
            parent[i] = i;
        }
        int ans = 0;
        int totalEdges = edges.size();
        int allowedEdges = 0;
        int i=0;
        while(allowedEdges < V-1){
            if(unionGroup(edges[i][0],edges[i][1],parent,rank)){
                allowedEdges++;
                ans += edges[i][2];
            }
            i++;
        }
        return ans;
    }
};

File: Graph/Kruskal/kru.py
from typing import List

class Solution:
    def findParent(self, a, parent):
        if parent[a] == a:
            return a
        parent[a] = self.findParent(parent[a], parent)
        return parent[a]

    def unionGroup(self, a, b, parent, rank):
        parA = self.findParent(a, parent)
        parB = self.findParent(b, parent)

        if parA == parB:
            return False

        rankA = rank[parA]
        rankB = rank[parB]

        if rankA >= rankB:
            parent[parB] = parA
            rank[parA] += 1
        else:
            parent[parA] = parB
            rank[parB] += 1

        return True

    def kruskalsMST(self, V: int, edges: List[List[int]]) -> int:
        edges.sort(key=lambda x: x[2])

        parent = [i for i in range(V)]
        rank = [1] * V

        ans = 0
        allowedEdges = 0
        i = 0

        while allowedEdges < V - 1:
            u, v, wt = edges[i]

            if self.unionGroup(u, v, parent, rank):
                allowedEdges += 1
                ans += wt

            i += 1

        return ans